### PR TITLE
Devel for future release (NUMA, Optane DC PMEM features)

### PR DESCRIPTION
**Pull request by ppalucki**
**Not ready for production use**

##### Bugs:
- [x] **PawelSz**: cpuset doesn't work for K8s (cannot write cpuset/mems), because cgroups deeper cpu hierrachy is used (works for Mesos)
- [x] **PawelP**: fix E2E tests,
- [x] **PawelSz**: make sure that swap is disable before using migrate_pages allocation (or warn about using it, it causes oom kills if there is not enough memory)
- [x] **PawelP**: ContainerSet aggregation should use "sum" by default for guage like metrics https://github.com/intel/workload-collocation-agent/issues/319
- [x] **PawelSz**: Fix inconsistency between unfified "mems" TaskResource in Mesos and Kubernetes (for Mesos it is expressed in MB) and document it (official docs; [aurora](http://aurora.apache.org/documentation/latest/reference/configuration/#resource-object)/[mesos](https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L1236)/[K8s](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory))
- [x] **Damian**: hardcoded "defaults" namespace for k8s task discovery [here](https://github.com/intel/workload-collocation-agent/blob/devel/wca/kubernetes.py#L109)
- [x] **PawelSz** Not supporting pod with not common path in cgroup (static pod)
- [x] **PawelSz** Wrong link to Intel RDT in installation docs.
- [x] **PawelSz** Fix checking sufficient privileges. ( WCA should work if no privileges are needed. )

##### Missing features:
- [x] **Pawel** make perf data (per task) gathering optional - 
- [x] **Pawel** rename action_delay to interval
- [x] **PawelSz**: Allocation API update (memory_migrate and other refer to new measurments.rst). (WCA),
- [x] **PawelP**: for all metrics with cpu level, attach additional socket label, to allow and simplify, aggregation,
- [x] **PawelP**: Missing metrics definition for dimm_count, dimm_total_size_bytes (with both labels ram and pmm) and memory_mode_size_bytes
- [x] **PawelSz**: Measurement API documentation. (WCA),
- [x] **PawelP**; refactor metric naming to use `platform__` and `task__prefix` and document this feature,
- [x] **PawelP**: unittests for wss collection,
- [x] **PawelP**: Improved metrics metadata - describe metrics properly 
- [x] **Damian**: replace migrate_pages binary with syscall
- [x] **PawelP**: derived metrics to make api complient with previous verision,
- [x] **PawelP/Marta**: additional metrics to cover UPI/QPI links (NUMA)
- [x] **PawelP/Marta**: Aggregate uncore perf counters (from IMC) - use hierachical levels introduced by Szymon (two levels: sockets, imc) 
- [x] **PawelP**: move memory migrate pages as proper allocator,
- [x] **PawelSz**: Make hardware/memory topology based on lshw and ipmctl optional
- [x] **PawelSz**: Expose numa distance in `Platform object` (based on `/sys/devices/system/node/node*/distance`)
- [x] **PawelSz**: Add auto-configuration to gather `platform.static_information`.
- [x] **Damian**: Memtier/mutilate based workload should be build from base centos7 image (2LM)
- [x] **Marta**: Rename `perf_pmu` to `perf_uncore` 
- [x] **PawelP**: Add derived metrics based on [this](https://github.com/intel/workload-collocation-agent/blob/devel/example/k8s_monitoring/prometheus/prometheus_rule.wca.yaml#L430) definitions (2LM)
- [x] **PawelSz**: Documentation for new/all metrics (`allocation.rst` and `metrics.rst`) (WCA)
- [x] **PawelSz**: Avoid redundant init variables in `runners` objects (ex. make global config object) (WCA)
- [x] **PawelSz**: Simplify API for plugins (instead of task_measurements/task_labels/task_resources - should have `{task: {measurements: {}, labels: {}}`) (WCA)

##### Docs:

- [x] **Szymon** numa allocator add README; add comment regarding huge pages (transparent huge pages)
- [x] **Szymon** Update docstring to follow one format, which is simple parsable (to create dynamic doc: API)
- [x] **PawelP**: Update metrics_source.rst how to disable for given metric source.
- [x] **PawelP**: check and fix all the links in frontpage docs and add link to metrics.rst and metrics_sources.rst

##### Polishing code:

- [x] **Szymon**: NumaAllocator and unittests, statictester/E2E,
- [x] **ALL**: ability to enable/disable collecting given feature (WSS, offcore metrics, numa metrics)
- [x] **Szymon**: refactor platform measurements to support hierarchical features 
- [x] **Damian**: Cleanup mutilate wrapper and wrapper code.
- [x] **PawelSz**: Refactor `are_privileges_sufficient` in `wca/security.py` to use needed only privileges.
- [x]  **PawelSz**: Require cgroup write access only when `AllocationRunner` is used,
- [x] **Damian** Add possibility connect to kubeapi outside pod
- [x] **Damian** Fix bug with send metric do Prometheus with wrong value

##### Testing and validation:

- [x] **PawelP Szymon**: Manually verify metrics are properly collected after refactoring and scale changes (check units).
- [x] **PawelP**: Move out the scheduler extender to new development branch (is untested), https://github.com/intel/workload-collocation-agent/pull/334/commits/0bcadff5fb3ab1195a330d347000a4666eb8d1ad
- [x] **Damian**: `StaticTester` integration tests, "cpuset, memset, memory_migrate"
- [x] **PawelP**: StaticTester cpus/mems and memory migrate, cgroup_migrate_memory,
- [x] **PawelSz**: unit test for WSS
- [x] **Marta**: unit tests for perf_pmu 
- [x] **Marta**: Remove eval based derived metrics  and factory abstraction 
- [x] **Damian**: E2E test for new Dockerfile in Kubernetes 
- [x] **Damian**: Move k8s workloads to workloads and cleanup workloads structure
- [x] **PawelSz**: Fix skipped unit tests (main, problem with logs) and enable config validation
- [x] **PawelSz**: Remove unit tests fixuture for KubernetesNode above 10k lines 
- [x] **Damian** Fix e2e on CI servers (Flake8, unit, update kernel)
- [x] **Szymon**: Fix recursive type definition.

##### Postponed after master memory_migrate (backlog) out of this PR

- verify 'examples/kubernetes' docs (rm duplicates) 
- consider if add test for migrate pages as syscall
- SpecJBB workload (2LM),
- remove unnecessary components (k8s manifests),
- Add headers about Copyrights to all files. 
- Restore workloads other than stress_ng and restore mutilate after removing it from e2e tests - "the docker image was not build with mutilate"

##### 1. Optane DC PMM support:

  - offcore/**uncore PMU** based metrics (memory bandwidth) 
  - **memory topology** available in `Platform object` - uses ipmctl and lshw
  - *customized workloads**: new way of deploying workloads using Kubernetes "kustomize",
  - new way of discovering tasks using **Kubernetes API**, 
  - multistage **Dockerfile** to build devel/packaged/thin versions of image,
  - Grafana **dashboards** for 2LM experiments,
  - k8s: **monitoring** - full deployment of full Kubernetes components including (node_exporter/kube-metric-state/prometheus operator and so on),
  - k8s: **scheduler extender** - scheduler extender to control placement decision,
  - **derived metrics:** new way of defining derived metrics (class based),
  - eval based **configurable derived metrics**,
  - approximate **working set size** (**WSS**) metric. based on "clean_refs and smaps" (overestimates for huge pages)
- new metric names with prefixes (**!BREAKING CHANGE!**)
  

##### 2. NUMA awareness:

 - new allocators:
   - improved CPUSet CPUS (more tests and fixed broken change detection)
   - explicit CPUSet MEMS 
   - enabling memory migration,
   - migrate_pages 
 - handling linux "list format" (encode/decode)
 - new `Platform` scoped measurements related to NUMA
 - new `Task scoped` measurements related to NUMA
 - new **NUMA Allocator** plugin

##### 3. Other

 - properly name the metrics according best practices from [Prometheus](https://prometheus.io/docs/practices/naming/) and automatically generate documentation,
 - deep hierarchical metrics "levels" like per task/per cpu/per node/per memory controller - perf event based metrics are now reporter in context of cpu,
 - fix the "missing measurements" exception - in case of missing exception the WCA will wait for next iteration
- new derived metrics,
- changes in scaling reports for hierarchical levels (TO BE REVISITED),
- common labels (like sockets, cpus etc. were moved to special metric called WCA_INFORMATION) https://github.com/intel/workload-collocation-agent/pull/328/commits/c97e691f7b4b95042c0a04da808e7b776f40a125
- removed default initial_task_cpu_assignment (label) need to be configured manually https://github.com/intel/workload-collocation-agent/pull/328/commits/7ec735c6fa6fd12e5fbafd46a04965d15067da26
- removed out of date dashboards and rules (https://github.com/intel/workload-collocation-agent/pull/334/commits/9a82ef542ebbab5e7a6f9c913c3b4824a60f7f9b) 
- remove scheduler extender PoC (https://github.com/intel/workload-collocation-agent/pull/334/commits/0bcadff5fb3ab1195a330d347000a4666eb8d1ad)

### CODE REVIEW

File: tests/e2e/demo_scenarios/common/wca_config_kubernetes_daemonset.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+runner: !AllocationRunner
+  rdt_enabled: True
+  node: !KubernetesNode
+    ssl: !SSL
+      server_verify: False
+      client_key_path: "CRT_PATH/apiserver-kubelet-client.key"
```
Comment (Damenus) [inline tests/e2e/demo_scenarios/common/wca_config_kubernetes_daemonset.yaml]: Maybe add comment that CRT_PATH will be replaced during e2e test

File: wca/storage_http.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+log = logging.getLogger(__name__)
+
+
+class ThreadingHTTPServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
+    daemon_threads = True
+
+
+@dataclass
+class HTTPStorage(Storage):
```
Comment (Damenus) [inline wca/storage_http.py]: SDL process allow us use wca as a server? Probably we need delete this.

File: wca/storage_http.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    port: int = 9100
+
+    def __post_init__(self):
+        self._message_counter = 0
+        self._output = ''
```
Comment (Damenus) [inline wca/storage_http.py]: There is no races?

File: examples/kubernetes/monitoring/fluentd/configurations/stress.conf [hunk 4]
```diff
@@ -0,0 +1,164 @@
+<filter kubernetes.var.log.containers.stress**>
+  @type parser
+  #@log_level trace
+  key_name log
+  reserve_data true
+  <parse>
+    ### BUG here: to second regexp is too broad and catches all the
```
Comment (Damenus) [inline examples/kubernetes/monitoring/fluentd/configurations/stress.conf]: Cleaned up comment in this file

File: examples/kubernetes/monitoring/fluentd/daemonset.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  template:
+    metadata:
+      name: fluentd
+      labels:
+        app: fluentd
+        version: v1
+    spec:
+      hostNetwork: true
+      serviceAccountName: fluentd
+      #tolerations:
```
Comment (Damenus) [inline examples/kubernetes/monitoring/fluentd/daemonset.yaml]: Deleted lines or comment?

**Comment by Damenus** [comment]
5645 added lines are from saved grafana dashboard

File: wca/runners/allocation.py [hunk 6]
```diff
@@ -233,6 +216,15 @@ def __init__(
 
         self._remove_all_resctrl_groups = remove_all_resctrl_groups
 
+        # Allocator need permission for writing to cgroups.
+        self._write_to_cgroup = True
```
Comment (ppalucki) [inline wca/runners/allocation.py]: @Creatone this value is not used at all - access to cgroups is not properly checked

File: wca/extra/README.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
+# WIP NUMA Allocator docs
```
Comment (Damenus) [inline wca/extra/README.md]: It is mandatory to completed.

File: .gitignore [hunk 8]
```diff
@@ -13,16 +16,20 @@ __pycache__/
 /gym_workloads-*
 .coverage
 /workloads/tensorflow-train/*.zip
```
[the remark this answers is not in the document]
Comment (ppalucki) [reply .gitignore]: not relevant

**Pull request merged**